core/image/jpeg/jpeg.odin

```diff
@@ -0,0 +1,927 @@
+			b = 0
+		}
+
+		z.code_buffer |= ((b << 56) >> u8(z.num_bits))
+		z.num_bits += 8
+		if z.num_bits > refill {
+			break
+		}
+	}
+}
+
+@(optimization_mode="favor_size", private="file")
+consume_bits_msb :: #force_inline proc(z: ^compress.Context_Memory_Input, width: u8) {
+	z.code_buffer <<= width
+	z.num_bits -= u64(width)
+}
+
+@(private="file")
+byte_align :: #force_inline proc(z: ^compress.Context_Memory_Input) {
+	skip := z.num_bits % 8
+	consume_bits_msb(z, cast(u8)skip)
+}
+
+@(optimization_mode="favor_size", private="file")
+peek_bits_msb :: #force_inline proc(z: ^compress.Context_Memory_Input, width: u8) -> u32 {
+	if z.num_bits < u64(width) {
+		refill_msb(z)
+	}
+	return u32((z.code_buffer &~ (max(u64) >> width)) >> (64 - width))
+}
+
+@(optimization_mode="favor_size", private="file")
+read_bits_msb :: #force_inline proc(z: ^compress.Context_Memory_Input, width: u8) -> u32 {
+	k := #force_inline peek_bits_msb(z, width)
+	#force_inline consume_bits_msb(z, width)
+	return k
+}
+
+load_from_bytes :: proc(data: []byte, options := Options{}, allocator := context.allocator) -> (img: ^Image, err: Error) {
+	ctx := &compress.Context_Memory_Input{
+		input_data = data,
+	}
+
+	img, err = load_from_context(ctx, options, allocator)
+	return img, err
+}
+
+@(private="file")
+get_symbol :: proc(ctx: ^$C, huffman_table: HuffmanTable) -> byte {
+	possible_code: u32 = 0
+
+	for i in 0..<HUFFMAN_MAX_BITS {
+		bit := read_bits_msb(ctx, 1)
+		possible_code = (possible_code << 1) | bit
+
+		for j := huffman_table.offsets[i]; j < huffman_table.offsets[i + 1]; j += 1 {
+			if possible_code == huffman_table.codes[j] {
+				return huffman_table.symbols[j]
+			}
+		}
+	}
+
+	return 0
+}
+
+load_from_context :: proc(ctx: ^$C, options := Options{}, allocator := context.allocator) -> (img: ^Image, err: Error) {
+	context.allocator = allocator
+	options := options
+
+	// Precalculate IDCT scaling factors
+	m0 := 2.0 * math.cos_f32(1.0 / 16.0 * 2.0 * math.PI)
+	m1 := 2.0 * math.cos_f32(2.0 / 16.0 * 2.0 * math.PI)
+	m3 := 2.0 * math.cos_f32(2.0 / 16.0 * 2.0 * math.PI)
+	m5 := 2.0 * math.cos_f32(3.0 / 16.0 * 2.0 * math.PI)
+	m2 := m0 - m5
+	m4 := m0 + m5
+
+	s0 := math.cos_f32(0.0 / 16.0 * math.PI) / math.sqrt_f32(8.0)
+	s1 := math.cos_f32(1.0 / 16.0 * math.PI) / 2.0
+	s2 := math.cos_f32(2.0 / 16.0 * math.PI) / 2.0
+	s3 := math.cos_f32(3.0 / 16.0 * math.PI) / 2.0
+	s4 := math.cos_f32(4.0 / 16.0 * math.PI) / 2.0
+	s5 := math.cos_f32(5.0 / 16.0 * math.PI) / 2.0
+	s6 := math.cos_f32(6.0 / 16.0 * math.PI) / 2.0
+	s7 := math.cos_f32(7.0 / 16.0 * math.PI) / 2.0
+
+	if .info in options {
+		options += {.return_metadata, .do_not_decompress_image}
+		options -= {.info}
+	}
+
+	if .return_header in options && .return_metadata in options {
+		options -= {.return_header}
+	}
+
+	first := compress.read_u8(ctx) or_return
+	soi := cast(image.JPEG_Marker)compress.read_u8(ctx) or_return
+	if first != 0xFF && soi != .SOI {
+		return img, .Invalid_Signature
+	}
+
+	img = new(Image)
```

Comment:
Can run out of memory, should handle the optional error from `new` here and return it as an error.